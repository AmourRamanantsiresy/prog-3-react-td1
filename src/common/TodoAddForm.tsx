import { ChangeEvent, FormEvent, useState } from 'react';
import { TI_INPUT_TODO_ADD } from '../tests/utils/selectors';
import TodoItem, { TodoStatus } from '../utils/models/Todo';

type TodoAddFormProps = {
  add: (todo: TodoItem) => void;
};

const TodoAddForm = (props: TodoAddFormProps) => {
  const [todoValue, setTodoValue] = useState<string>('');
  const { add } = props;

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (todoValue.length > 0) {
      add({ id: new Date().getTime(), status: TodoStatus.TODO, value: todoValue });
      setTodoValue('');
    }
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setTodoValue(event.target.value);
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        data-testid={TI_INPUT_TODO_ADD}
        placeholder='Add task'
        value={todoValue}
        type='text'
        onChange={handleChange}
      />
    </form>
  );
};

export default TodoAddForm;
