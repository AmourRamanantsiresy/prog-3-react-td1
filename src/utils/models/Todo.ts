export enum TodoStatus {
  TODO = 'todo',
  DONE = 'done',
}

type Todo = {
  value: string;
  id: number;
  status: TodoStatus;
};

export default Todo;
