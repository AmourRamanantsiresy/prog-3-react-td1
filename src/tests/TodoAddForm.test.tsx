import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { create } from 'react-test-renderer';
import TodoAddForm from '../common/TodoAddForm';
import Todo, { TodoStatus } from '../utils/models/Todo';

describe('Test TodoAddForm', () => {
  test('Architecture', () => {
    const tree = create(
      <TodoAddForm
        add={todo => {
          console.log(todo);
        }}
        key='todoAddForm-1'
      />
    );

    expect(tree).toMatchSnapshot();
  });

  test('Add new todo', () => {
    const addTodo: jest.Mock<Todo> = jest.fn(todo => todo);
    const valueOfNewTodo: string = 'This is a new todo for test';
    const idBeforeNewTodo: number = new Date().getTime();

    render(<TodoAddForm add={addTodo} key='todoAddForm-1' />);

    const addTodoInput = screen.getByPlaceholderText('Add task');

    expect(addTodoInput).toBeTruthy();
    userEvent.type(addTodoInput, `${valueOfNewTodo}{Enter}`);
    expect(addTodo.mock.results).not.toBeNull();
    const { id, status, value }: Todo = addTodo.mock.results[0].value;

    expect(status).toEqual(TodoStatus.TODO);
    expect(value).toEqual(valueOfNewTodo);
    expect(idBeforeNewTodo).toBeLessThanOrEqual(id);
    expect(id).toBeLessThanOrEqual(new Date().getTime());
  });
});
