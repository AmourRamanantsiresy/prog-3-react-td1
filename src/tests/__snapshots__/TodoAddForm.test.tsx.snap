// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoAddForm Architecture 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    data-testid="input-todo-add-id"
    onChange={[Function]}
    placeholder="Add task"
    type="text"
    value=""
  />
</form>
`;
